bin/cloudtrail_handler.py

```diff
@@ -39,36 +39,45 @@ def __init__(self, config, s3, query_dao):
         self.query_dao = query_dao
 
     def process_log(self, event):
-        s3_record = event['Records'][0]['s3']
-        bucket = s3_record['bucket']['name']
-        key = s3_record['object']['key']
+        s3_record = event["Records"][0]["s3"]
+        bucket = s3_record["bucket"]["name"]
+        key = s3_record["object"]["key"]
         object = self.s3.Object(bucket, key).get()
-        stream = BytesIO(object['Body'].read())
-        with gzip.GzipFile(None, 'rb', fileobj=stream) as file:
+        stream = BytesIO(object["Body"].read())
+        with gzip.GzipFile(None, "rb", fileobj=stream) as file:
             try:
                 data = json.load(file)
             except json.JSONDecodeError:
-                logger.warning(f'Not a valid cloudtrail json file {bucket}/{key}', exc_info=True)
+                logger.warning(f"Not a valid cloudtrail json file {bucket}/{key}", exc_info=True)
                 return
-            for record in data.get('Records', []):
-                if record.get('eventName') == 'StartQueryExecution':
+            for record in data.get("Records", []):
+                if record.get("eventName") == "StartQueryExecution":
                     self.process_query_record(record)
 
     def process_query_record(self, record):
-        if record.get('responseElements') and record.get('userIdentity'):
-            time = datetime.strptime(record['eventTime'], CLOUDTRAIL_TIME_FORMAT).replace(
-                tzinfo=timezone.utc)
-            if record['userIdentity']['type'] == 'IAMUser':
-                inferred_executing_user = record['userIdentity']['userName']
-            if record['userIdentity']['type'] == 'AssumedRole':
-                inferred_executing_user = 'API'
+        if record.get("responseElements") and record.get("userIdentity"):
+            time = datetime.strptime(record["eventTime"], CLOUDTRAIL_TIME_FORMAT).replace(tzinfo=timezone.utc)
+            inferred_executing_user = "API"
+            if record["userIdentity"]["type"] == "IAMUser":
+                inferred_executing_user = record["userIdentity"]["userName"]
+            if record["userIdentity"]["type"] == "AssumedRole":
+                if "arn" in record["userIdentity"]:
+                    arn = record["userIdentity"]["arn"]
+                    arn_parts = arn.split("/")
+                    if len(arn_parts) >= 3:
```

Comment:
I'm not a fan of hardcoding/striping/indexing strings, but this seems to be the easiest/clearest way to achieve what we need;)